### PR TITLE
Added solution to SSSP Negative problem

### PR DESCRIPTION
**Pull request by fbrunodr**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

File: content/6_Advanced/SPneg.mdx [hunk 1]
```diff
@@ -43,7 +43,100 @@ frequency: 0
...
-<IncompleteSection />
+Although this is a Single Source Shortest Paths problem, we can not use the known
+Dijkstra's algorithm, because we have negative weights on edges. An alternative is
+to use the Bellman-Ford algorithm. The algorithm first considers all the paths
+which use 1 edge. Then it calculates all the paths with at most 2 edges, and so
+on. If the graph has no negative cycle, then the shortest path between the source
+and any other vertice should have at most V-1 edges, where V stands for the number
+of vertices in the graph. Because of that, the algorithm iterates through at most
+all edges V-1 times. Hence, it runs in O(E * V).
+
```
Comment (ryanchou-dev) [inline content/6_Advanced/SPneg.mdx:54]: could you use latex?

File: content/6_Advanced/SPneg.mdx [hunk 2]
```diff
@@ -43,7 +43,100 @@ frequency: 0
...
+If the graph has a negative cycle, we can detect a vertice in this cycle by running
+another relaxation. In this problem, belonging to a negative cycle means the
+distance to that point is minus infinity. Notice that all points that are reachable
+from those will also have minus infinity cost. In the solution below, we detect
+all the negative cycles and save the point from which we detected the cycle. We then
+do a bfs with those points as sources.
```
Comment (ryanchou-dev) [inline content/6_Advanced/SPneg.mdx:60]:
```suggestion
If the graph has a negative cycle, we can detect a vertice in this cycle by running
another relaxation. In this problem, belonging to a negative cycle means the
distance to that point is negative infinity. Notice that all points that are reachable
from those will also have minus infinity cost. In the solution below, we detect
all the negative cycles and save the point from which we detected the cycle. We then
do a [BFS](/silver/graph-traversal) with those points as sources.
```

File: content/6_Advanced/SPneg.mdx [hunk 3]
```diff
@@ -43,7 +43,100 @@ frequency: 0
...
+			}
+		}
+
+		for (int i = 0; i < q; i++) {
+			int target;
+			cin >> target;
+			if (is_negative_inf[target]) {
+				cout << "-Infinity\n";
+			} else if (dist[target] == INF) {
+				cout << "Impossible\n";
```
Comment (ryanchou-dev) [inline content/6_Advanced/SPneg.mdx:126]: keep it consistent :D
```suggestion
				cout << "-Infinity" << '\n';
			} else if (dist[target] == INF) {
				cout << "Impossible" << '\n';
```

File: content/6_Advanced/SPneg.mdx [hunk 4]
```diff
@@ -43,7 +43,100 @@ frequency: 0
...
+do a bfs with those points as sources.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+using ii = pair<int, int>;
```
Comment (ryanchou-dev) [inline content/6_Advanced/SPneg.mdx:69]: (optional)

remove this? (you only use this once)

File: content/6_Advanced/SPneg.mdx [hunk 5]
```diff
@@ -43,7 +43,100 @@ frequency: 0
...
+
+const int INF = 1e9;
+
+int main() {
+	int n, m, q, s;
+
+	cin >> n >> m >> q >> s;
+	while (!(n == 0 && m == 0 && q == 0 && s == 0)) {
```
Comment (ryanchou-dev) [inline content/6_Advanced/SPneg.mdx:77]:
```suggestion
	int n, m, q, s;
	cin >> n >> m >> q >> s;
	
	while (!(n == 0 && m == 0 && q == 0 && s == 0)) {
```

**Review by ryanchou-dev — COMMENTED** [review]
awesome!

File: content/6_Advanced/SPneg.mdx [hunk 6]
```diff
@@ -43,7 +43,99 @@ frequency: 0
...
+#include <bits/stdc++.h>
+using namespace std;
+
+const int INF = 1e9;
+
+int main() {
+	int n, m, q, s;
+	cin >> n >> m >> q >> s;
+
+	while (!(n == 0 && m == 0 && q == 0 && s == 0)) {
```
Comment (SansPapyrus683) [inline content/6_Advanced/SPneg.mdx:76]: what's the while loop for

File: content/6_Advanced/SPneg.mdx [hunk 7]
```diff
@@ -43,7 +43,99 @@ frequency: 0
...
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int INF = 1e9;
```
Comment (SansPapyrus683) [inline content/6_Advanced/SPneg.mdx:70]: why not INT32_MAX

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/6_Advanced/SPneg.mdx (hunk 6 above)
Comment (fbrunodr) [reply content/6_Advanced/SPneg.mdx:76]: See problem statement

File: content/6_Advanced/SPneg.mdx (hunk 7 above)
Comment (fbrunodr) [reply content/6_Advanced/SPneg.mdx:70]: Avoid overflow when performing addition. For this implementation in particular it does not make a difference, but it is usually helpful that INF + INF does not overflow.

Comment (SansPapyrus683) [reply content/6_Advanced/SPneg.mdx:70]: so are you doing inf + inf’s here or not

Comment (fbrunodr) [reply content/6_Advanced/SPneg.mdx:70]: Not doing

Comment (SansPapyrus683) [reply content/6_Advanced/SPneg.mdx:70]: then just use int32_max

Comment (fbrunodr) [reply content/6_Advanced/SPneg.mdx:70]: Ok

**Review by SansPapyrus683 — APPROVED** [review]

**Comment by SansPapyrus683** [comment]
my guy i alr approved

**Review by ryanchou-dev — APPROVED** [review]
good

**Pull request merged**